src/olympia/files/models.py

```diff
@@ -378,40 +378,53 @@ def get_localepicker(self):
     def webext_permissions(self):
         """Return permissions that should be displayed, with descriptions, in
         defined order:
-        1) match all urls (e.g. <all-urls>)
-        2) known permissions, in constants order (alphabetically),
-        3) match urls for sites
+        1) Either the match all permission, if present (e.g. <all-urls>), or
+           match urls for sites (<all-urls> takes preference over match urls)
+        2) nativeMessaging permission, if present
+        3) other known permissions in alphabetical order
         """
         knowns = list(WebextPermissionDescription.objects.filter(
             name__in=self.webext_permissions_list).iterator())
-
-        match_all, urls = [], []
+        # Move nativeMessaging permission to front of the list, if present
+        for index, perm in enumerate(knowns):
```

Comment:
it's not possible - the urls need to be grouped if there are more than one, and then added to the star, and any of <all_urls> could be in any position too.  Plus the first loop, here, is for the permission descriptions from the database, whereas the second loop is for the permissions from the manifest, which could be in any order.
I can't see an algorithm with less than 2 a) for loops b) sorts (which internally involve loops) or c) `x in list` lookups, which also are similarly expensive.